### PR TITLE
add singular resource names to discovery

### PR DESCRIPTION
**Pull request by deads2k**
Adds the singular resource name to our resource for discovery.  This is something we've discussed to remove our pseudo-pluralization library which is unreliable even for english and really has no hope of properly handling other languages or variations we can expect from TPRs and aggregated API servers.

This pull simply adds the information to discovery, it doesn't not re-wire any RESTMappers.

@kubernetes/sig-cli-misc  @kubernetes/sig-apimachinery-misc @kubernetes/api-review

```release-note
API resource discovery now includes the `singularName` used to refer to the resource.
```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 1]
```diff
@@ -682,8 +682,10 @@ type GroupVersionForDiscovery struct {
 
 // APIResource specifies the name of a resource and whether it is namespaced.
 type APIResource struct {
-	// name is the name of the resource.
+	// name is the plural name of the resource.
 	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// singular is the singular name of the resource.  This allows clients to handle plural and singular opaquely.
+	Singular string `json:"singular" protobuf:"bytes,6,opt,name=singular"`
```
Comment (lavalamp) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: I prefer "SingularName"

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 2]
```diff
@@ -682,8 +682,10 @@ type GroupVersionForDiscovery struct {
 
 // APIResource specifies the name of a resource and whether it is namespaced.
 type APIResource struct {
-	// name is the name of the resource.
+	// name is the plural name of the resource.
 	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// singular is the singular name of the resource.  This allows clients to handle plural and singular opaquely.
```
Comment (lavalamp) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: Add to documentation where the plural name is used vs where the singular name is used.

**Comment by lavalamp** [comment]
API change lgtm if you fix my comments.

Are you planning on doing a second PR or adding the implementation to this one?

**Comment by deads2k** [comment]
> Are you planning on doing a second PR or adding the implementation to this one?

Second PR.  I may even see if I can con @p0lyn0mial into trying it :)

**Comment by lavalamp** [comment]
@mbohlool is there a convenient place in OpenAPI to list this sort of metadata?

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go (hunk 1 above)
Comment (sttts) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: +1

**Comment by sttts** [comment]
lgtm with `SingularName` as field name.

Let's also keep this in mind for TPRs. Now the kind and resource are linked, i.e. either singular or plural.

**Comment by deads2k** [comment]
link	@k8s-bot cvm gce e2e test this

**Comment by deads2k** [comment]
@k8s-bot cvm gce e2e test this

**Comment by deads2k** [comment]
Name updated.

**Comment by mbohlool** [comment]
Generated and non-generated code are both in one commit that make it hard to review.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 3]
```diff
@@ -682,8 +682,12 @@ type GroupVersionForDiscovery struct {
 
 // APIResource specifies the name of a resource and whether it is namespaced.
 type APIResource struct {
-	// name is the name of the resource.
+	// name is the plural name of the resource.
 	Name string `json:"name" protobuf:"bytes,1,opt,name=name"`
+	// singularName is the singular name of the resource.  This allows clients to handle plural and singular opaquely.
+	// The singularName is more correct for reporting status on a single item and both singular and plural are allowed
+	// from the kubectl CLI interface.
+	SingularName string `json:"singularName" protobuf:"bytes,6,opt,name=singularName"`
```
Comment (deads2k) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go]: @mbohlool only line of meat here.

**Comment by mbohlool** [comment]
Yep, I saw 54 files changes and was looking for some kind of implementation for this :). In future if we split the commits, it make it really easy to review.

Regarding openapi, the implementation should also add a tag to openapi spec of that model. e.g., adding a `"x-kubernetes-plural-name": "pods"` to `pod` spec. To add a tag to a type, you can follow the example here (`GetDefinitionName` function): https://github.com/kubernetes/apiserver/blob/aa75ef0b36d8ca79160d9f8f3e8a9b926ee30c1b/pkg/endpoints/openapi/openapi.go#L163

**Comment by deads2k** [comment]
@k8s-bot kops aws e2e test this
@k8s-bot bazel test this

**Pull request merged**